benchmarks/ALCE/ELI5/eli5.py

```diff
@@ -0,0 +1,152 @@
+import json
+import os
+import time
+
+from datasets import Dataset
+from tqdm import tqdm
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+import rageval as rl
+from rageval.models.openai import OpenAILLM
+from utils import make_demo, create_eli5_eval_dataset
+
+
+class ELI5:
+
+    def __init__(self, args):
+        self.args = args
+
+        print("-" * 10 + "Loading model" + "-" * 10)
+        if "gpt-3.5-turbo" in self.args.model:
+            os.environ["OPENAI_API_KEY"] = self.args.api_key
+            self.model = OpenAILLM(
+                model=self.args.model,
+                _api_key_env_var="OPENAI_API_KEY",
+                max_tokens=self.args.max_tokens,
+                temperature=self.args.temperature,
+                top_p=self.args.top_p
+            )
+        else:
+            self.tokenizer = AutoTokenizer.from_pretrained(self.args.model, use_fast=False)
+            self.model = AutoModelForCausalLM.from_pretrained(
+                self.args.model,
+                device_map='auto'
+            )
+        print("-" * 10 + "Finish loading model" + "-" * 10)
+
+        print("-" * 10 + "Loading dataset" + "-" * 10)
+        if self.args.dataset == "bm25":
+            self.eval_data = json.load(
+                open(self.args.cache_path + "/datasets/ALCE-data/eli5_eval_bm25_top100.json", "r")
+            )
+        elif self.args.dataset == "oracle":
+            self.eval_data = json.load(
+                open(self.args.cache_path + "/datasets/ALCE-data/eli5_eval_bm25_top100_reranked_oracle.json", "r")
+            )
+        else:
+            raise ValueError("Don't support such dataset.")
+
+        self.prompt = json.load(
+            open("benchmarks/ALCE/ELI5/prompts/eli5_prompt.json", "r")
+        )
+        print("-" * 10 + "Finish loading dataset" + "-" * 10)
+
+        model_name = self.args.model
+        if "/" in model_name:
+            model_name = model_name.split("/")[-1]
+        self.name = f"eli5-{self.args.dataset}-{model_name}-{self.args.method}-shot{self.args.shot}-ndoc{self.args.ndoc}"
+        self.result_path = ".rageval/results/" + self.name + ".json"
+
+    def generate(self):
+        head_prompt = ""
```

Comment:
`generate` could be implemented elsewhere, and then simpy pass the results to the `evaluate` function, rather than coupling with specific input args. With the current implementations, If the model changes, such as accessing a model via an API, even the `init` part should be modified.

Reply:
> `generate` could be implemented elsewhere, and then simpy pass the results to the `evaluate` function, rather than coupling with specific input args. With the current implementations, If the model changes, such as accessing a model via an API, even the `init` part should be modified.

It means to split the `generate` here into two steps: `preprocess` and `generate`. Among them, `preprocess` is implemented based on different data sets, and `generate` is abstracted into a base class function?

Reply:
> > `generate` could be implemented elsewhere, and then simpy pass the results to the `evaluate` function, rather than coupling with specific input args. With the current implementations, If the model changes, such as accessing a model via an API, even the `init` part should be modified.
> 
> Does it mean that the `generate` here can be split into two steps: `preprocess` and `generate`? Among them, `preprocess` is implemented based on different data sets, and `generate` is abstracted into a base class function?

In my view, `generation` should not be tightly coupling with a benchmark, users might prefer their own implementations. When running a benchmark, we could simply define the required dataset format and use the input dataset along with model outputs for evaluation. In other words, a benchmark is a set of metrics accompanied by an example of how to generate and format model output.